### PR TITLE
Support 16-bit pointers as well as i/usize

### PR DESCRIPTION
**Pull request by shepmaster**
I'm opening this pull request to get some feedback from the community.

Although Rust doesn't support any platforms with a native 16-bit pointer at the moment, the [AVR-Rust](https://github.com/avr-rust/rust) fork is working towards that goal. Keeping this forked logic up-to-date with the changes in master has been onerous so I'd like to merge these changes so that they get carried along when refactoring happens. I do not believe this should increase the maintenance burden.

This is based on the original work of Dylan McKay (@dylanmckay).

### CODE REVIEW

**Comment by alexcrichton** [comment]
This all seems pretty straightforward and easy to land, but in terms of keeping this up to date is there a way that we could add a simple regression test? There's a lot more places than I would have expected that need to be updated for 16-bit pointers, and ensuring that any new location stays up to date may be difficult.

**Comment by shepmaster** [comment]
> is there a way that we could add a simple regression test

I'm 100% behind this, but I'm pretty stumped on how to do it. It seems like I need to be able to say "hey, this platform has 16-bit pointers", but there aren't any that do so (for now).

Do you think there's a way I could use a target description JSON file, specify `"target-pointer-width": "16",` and have it work on (say) x86? Then I could just do a bunch of math-like stuff, each corresponding to one of the changes being made?

> a lot more places than I would have expected

You're telling me! 🌴 Unfortunately, most of them were found The Hard Way — by finding a compiler ICE building the compiler or my end code. It's a shame that there wasn't more "hey you forgot this enum value here", but most cases (reasonably) have a default value case.

**Comment by nagisa** [comment]
The way I remember, on 16-bit systems you usually would have a distinction between far (segment+ptr) and near pointers, which seems to not be addressed here. What gives?

**Comment by shepmaster** [comment]
@nagisa I have heard tales of Ye Olde Days of `near` and `far` pointers, but I'm not sure that AVR has need for it, as there isn't a segmented memory. There's [some references to it](http://www.atmel.com/webdoc/AVRLibcReferenceManual/ch20s05s01.html) when one needs to access > 64K of memory, but those seem to just be "regular" 32-bit values.

**Comment by dylanmckay** [comment]
@shepmaster is right in saying AVR doesn't use segmented memory. Normal pointers are the usual 16 bits. There are a few AVR chips with a bit more memory that 16 bits can address - to access memory higher than this, there's a special IO register for which you must load the other address bits into first. It still works as an ordinary pointer, but spread over two registers.

**Comment by shepmaster** [comment]
> a way I could use a target description JSON file

Mmm, that doesn't seem to work because it ultimately looks like cross compiling and there's no libcore or libstd for that made up platform. I'm open to any suggestions!

**Comment by shepmaster** [comment]
And If I try to follow the lead of the target-specification tests and eschew libstd and libcore, I end up triggering an LLVM assertion:

```
LLVM ERROR: Cannot select: 0x10e629000: i16 = FrameIndex<0>
In function: _ZN4main5_main17h49c054cd6e240b5cE
```

**Comment by shepmaster** [comment]
I *can* get a tiny bit of test with this:

```rust
let _a: usize = 0x1_0000; // literal out of range for usize
```

Which indicates that, yes, 16-bit pointers only take up 16-bits. I'm not sure of the value of the test, but I'll continue poking at it. I'd still appreciate suggestions that anyone has for useful tests.

**Comment by alexcrichton** [comment]
Hm yeah I suspect that our main recourse for testing a patch like this would be adding automation which generated at least libcore for a 16-bit target.

Also, forgot to do this earlier, but cc @rust-lang/compiler and @rust-lang/tools

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -178,6 +179,7 @@ impl TargetDataLayout {
 
     pub fn ptr_sized_integer(&self) -> Integer {
         match self.pointer_size.bits() {
+            16 => I16,
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Neat.

**Comment by alexcrichton** [comment]
The tools team discussed this PR during triage yesterday and the conclusion was that from our perspective this seems fine. This would basically just be expanding our tier 3 platforms, which we regularly take PRs for, so no sweat on that front!

**Comment by shepmaster** [comment]
@alexcrichton that's great! I found one more tricky spot that needs to be updated, I can rebase and squash that to this commit and it should be good-to-merge then!

**Comment by shepmaster** [comment]
@alexcrichton was there any comment one way or the other about tests that would need to be added before the merge?

**Comment by alexcrichton** [comment]
At least from a tooling perspective we were ok not having a bunch of rigorous tests or anything, none of our tier 3 platforms do either :)

For merging, though, this cuts across a few other teams as well, so at least @rust-lang/compiler will probably want to sign off as well

**Comment by shepmaster** [comment]
I've fixed up the one extra spot I've found and rebased for good measure.

**Comment by nikomatsakis** [comment]
I gave it a quick read. Looks like a fairly straight-forward patch; I've not thought deeply about what else might be needed to support 16 bit platforms, but this at least doesn't look like it would harm 32 or 64 bit. 👍

**Comment by alexcrichton** [comment]
The libs team discussed this briefly at yesterday's triage and the conclusion was that from that perspective we're ok along the lines of "let's see how this plays out without bending over backwards too much"

**Comment by pnkfelix** [comment]
I'm cool with us landing this.  Might give us more street cred with micro controller community

**Comment by shepmaster** [comment]
Thanks for the positive reviews all around! What other teams remain to sign off?

**Comment by alexcrichton** [comment]
Judging by the area of change, I'd expect that at least the compiler team should weigh in. If they'd like to escalate, however, we can discuss at the core meeting.

**Comment by pnkfelix** [comment]
@nikomatsakis we did discuss at compiler meeting (twice I think,  right?)

We didn't really attempt to form any solid consensus (we almost never have a formal quorum anyway), but no one present *objected*...

**Comment by nikomatsakis** [comment]
@pnkfelix indeed, we did discuss, and everybody was more-or-less in favor of doing this on an experimental basis.

**Comment by Aatch** [comment]
I think the conclusion is that this is ok.

@bors r+

**Pull request merged**